### PR TITLE
Transactional password change

### PR DESCRIPTION
**Pull request by baptiste-grob**
The password change operation now has the following flow: 
1. Check password locally.
    - Abort if the password-computed root key does not match the local root key
2. Create and set a new root key
3. Create and set a new default items key
4. Sync
5. Change the password on the server

If step 4 or 5 fails: 
1. Set the root key back to its previous value
2. Delete the new items key
3. Set the default items key back to its previous value
4. If step 4 was successful, sync
    - If this sync fails, we could have a client/server mismatch and logging out would essentially make the account unusable. An alternative would be to wait for the server password change to succeed before setting a new default items key, however that would not work in the case of a protocol upgrade.

### CODE REVIEW

**Comment by baptiste-grob** [comment]
Remaining tests to write:
- [] can still log in and decrypt items even if the network goes down right after a server protocol upgrade
- [] rolls back the local protocol upgrade if the server responds with an error

The application code is ready to review however.

**Comment by baptiste-grob** [comment]
Ready for review 👍

File: lib/application.ts [hunk 1]
```diff
@@ -1050,31 +1055,56 @@ export class SNApplication {
...
+      newRootKey,
+      newKeyParams,
+      rollback: rollbackPasswordChange
+    } = changePasswordResult!;
+
+    /** Sync the newly created items key. Roll back on failure */
+    await this.syncService!.sync({ awaitAll: true });
+    const itemsKeyWasSynced = this.protocolService!.getDefaultItemsKey()!.updated_at.getTime() > 0;
+    if (!itemsKeyWasSynced) {
+      await rollbackPasswordChange();
```
[the remark this answers is not in the document]
Comment (baptiste-grob) [reply lib/application.ts:1088]: I think you're pointing out a hidden assumption about my code, which is that if `itemsKeyWasSynced` is false, then no sync has occurred. Thus we would only need to do a local rollback and the local state would equal the server's state. I must ask then, if the following scenario can occur:
- change: `previousItemsKey.isDefault = false` _was_ synced to the server
- change: `newItemsKey.isDefault = true` _was not_ synced to the server
If my check doesn't cover this case then i should change it, for example by checking the updated_at on the previous items key.

File: lib/application.ts [hunk 2]
```diff
@@ -1050,31 +1055,56 @@ export class SNApplication {
     currentPassword: string,
     newPassword: string,
     passcode?: string
-  ) {
+  ): Promise<{ error?: Error }> {
+    if (newPassword.length < MINIMUM_PASSWORD_LENGTH) {
```
[the remark this answers is not in the document]
Comment (baptiste-grob) [reply lib/application.ts]: See the code update for my proposed solution

File: lib/application.ts (hunk 1 above)
[the remark this answers is not in the document]
Comment (mobitar) [reply lib/application.ts:1092]: I suppose that's true, but we don't know exactly why the sync failed. Perhaps the sync went through on the server and data was changed, but the server returned an exception nonetheless, or the request was dropped? In that case, wouldn't that change need to be undone?

[the remark this answers is not in the document]
Comment (baptiste-grob) [reply lib/application.ts:1092]: Oh right, that's a good point.

[the remark this answers is not in the document]
Comment (baptiste-grob) [reply lib/application.ts:1092]: Updated

**Comment by baptiste-grob** [comment]
Heads up @johnny243! This PR adds the Sinon.JS dependency which you're already using on your branch I believe, so you may have some conflicts as a result of me merging the PR in.

**Comment by johnny243** [comment]
I'll keep it in mind, thanks!

**Pull request merged**